### PR TITLE
Lesson20

### PR DESCRIPTION
**Pull request by pandaUbuntu**

### CODE REVIEW

File: app/code/Oleksandrr/Chat/Controller/Message/Messages.php [hunk 1]
```diff
@@ -19,35 +20,56 @@ class Messages extends \Magento\Framework\App\Action\Action implements
...
-            $messageCollection = $this->messageCollectionFactory->create();
-            $messageCollection
-                ->addFieldToSelect('author_name')
-                ->addFieldToSelect('message')
-                ->addFieldToSelect('created_at')
-                ->setPageSize(10);
-            $collection = $messageCollection->getData();
-            $message = 'Post count: ' . count($collection);
+            if ($this->userSession->isLoggedIn()) {
+                $messageCollection = $this->messageCollectionFactory->create()->addAuthorFilter((int) $this->userSession->getId());
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Messages.php]: @pandaUbuntu Should filter collection by chat hash, not customerId - otherwise you do not get admin answers. Collection is anyway filtered by chat hash - both for guests and registered customers.

By the way, use the word `customer` (e.g. buyer) for storefront visitors, not `user`. People with access to Admin Panel are `users`

File: app/code/Oleksandrr/Chat/Controller/Message/Messages.php [hunk 2]
```diff
@@ -19,35 +20,56 @@ class Messages extends \Magento\Framework\App\Action\Action implements
...
+            } else {
+                if ($chatHash = $this->userSession->getChatHash()) {
+                    $messageCollection = $this->messageCollectionFactory->create()->addChatHashFilter($chatHash);
+                }
+            }
+
+            if (isset($messageCollection)) {
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Messages.php]: @pandaUbuntu This is always `true` because $messageCollection is object.

File: app/code/Oleksandrr/Chat/Controller/Message/Messages.php [hunk 3]
```diff
@@ -19,35 +20,56 @@ class Messages extends \Magento\Framework\App\Action\Action implements
...
+                /** @var Message $message */
+                foreach ($messageCollection as $message) {
+                    $collection[] = [
+                        'author_name' => $message->getAuthorName(),
+                        'message' => $message->getMessage(),
+                        'created_at' => $message->getCreatedAt(),
+                    ];
+                }
+
+                $message = 'Post count: ' . count($collection);
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Messages.php]: @pandaUbuntu `$collection` is \Traversable, but better to use `$collection->count()` instead.

File: app/code/Oleksandrr/Chat/Controller/Message/Save.php [hunk 4]
```diff
@@ -29,50 +30,81 @@ class Save extends \Magento\Framework\App\Action\Action implements
...
      */
     public function __construct(
         \Magento\Framework\App\Action\Context $context,
         \Oleksandrr\Chat\Model\MessageFactory $messageFactory,
         \Magento\Store\Model\StoreManagerInterface $storeManager,
         \Psr\Log\LoggerInterface $logger,
-        \Oleksandrr\Chat\Model\ResourceModel\Message $messageResource
+        \Oleksandrr\Chat\Model\ResourceModel\Message $messageResource,
+        \Magento\Customer\Model\Session $userSession,
+        \Magento\Framework\Data\Form\FormKey\Validator $formKeyValidator
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Save.php:60]: @pandaUbuntu Parameter is not annotated

File: app/code/Oleksandrr/Chat/Controller/Message/Save.php [hunk 5]
```diff
@@ -29,50 +30,81 @@ class Save extends \Magento\Framework\App\Action\Action implements
...
+            if (!$this->formKeyValidator->validate($request)) {
+                throw new LocalizedException(__('Validation Failed'));
+            }
+
+            $userId = (int) $this->userSession->getId();
             $websiteId = (int) $this->storeManager->getWebsite()->getId();
 
             /** @var \Oleksandrr\Chat\Model\Message $message */
             $message = $this->messageFactory->create();
+            $date = new \DateTime();
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Save.php]: @pandaUbuntu Don't use native PHP classes to work with date/time - they do not handle store locale settings and thus date may be saved incorrectly, because this will most likely be a GMT date. Check \Magento\Framework\Stdlib\DateTime for classes that work with date/time.

File: app/code/Oleksandrr/Chat/Controller/Message/Save.php [hunk 6]
```diff
@@ -29,50 +30,81 @@ class Save extends \Magento\Framework\App\Action\Action implements
...
+
+            if (!$hash = $this->userSession->getChatHash()) {
+                $hash = md5($date->getTimestamp() . $request->getParam('user_message'));
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Save.php]: @pandaUbuntu This code is vulnerable, because it is easy to hack chat hashes. This is not really critical in this particular case though. Anyway, it is better to generate hash based on `uniqid()` or some other functions instead of dealing with dates and message text.

File: app/code/Oleksandrr/Chat/Controller/Message/Save.php [hunk 7]
```diff
@@ -29,50 +30,81 @@ class Save extends \Magento\Framework\App\Action\Action implements
...
+                $this->userSession->setChatHash($hash);
+            }
 
-            $message->setAuthorId(1)
+            $message
+                ->setAuthorId($userId ?: 0)
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Save.php]: @pandaUbuntu Just leave NULL for guests

File: app/code/Oleksandrr/Chat/Controller/Message/Save.php [hunk 8]
```diff
@@ -29,50 +30,81 @@ class Save extends \Magento\Framework\App\Action\Action implements
...
                 ->setAuthorType(1)
                 ->setWebsiteId($websiteId)
                 ->setMessage($request->getParam('user_message') ?: '')
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Save.php]: @pandaUbuntu 
1) This code has better equivalent because the method `getParam` takes two arguments: `$request->getParam('user_message', '')`
2) Empty message is a problem and this should trigger error.

File: app/code/Oleksandrr/Chat/Controller/Message/Save.php [hunk 9]
```diff
@@ -29,50 +30,81 @@ class Save extends \Magento\Framework\App\Action\Action implements
...
+
+            if (!$hash = $this->userSession->getChatHash()) {
+                $hash = md5($date->getTimestamp() . $request->getParam('user_message'));
+                $this->userSession->setChatHash($hash);
+            }
 
-            $message->setAuthorId(1)
+            $message
+                ->setAuthorId($userId ?: 0)
                 ->setAuthorType(1)
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Save.php]: @pandaUbuntu Should move author types to some constant in the Model

File: app/code/Oleksandrr/Chat/Observer/UserLogin.php [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        $this->userSession = $userSession;
+    }
+
+    public function execute(\Magento\Framework\Event\Observer $observer)
+    {
+        try {
+            if ($chatHash = $this->userSession->getChatHash()) {
+                $user = $observer->getEvent()->getCustomer();
+                $messageCollection = $this->messageCollectionFactory->create();
+                $messageCollection->addChatHashFilter($chatHash)->addAuthorTypeFilter(1);
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Observer/UserLogin.php]: @pandaUbuntu Admin responses to guest will not be updated in this case

File: app/code/Oleksandrr/Chat/Observer/UserLogin.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+                /** @var Message $message */
+                foreach ($messageCollection as $message) {
+                    if ((int) $message->getAuthorId() === 0) {
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Observer/UserLogin.php]: @pandaUbuntu `if not` would be enough in the code like this

File: app/code/Oleksandrr/Chat/Controller/Message/Messages.php [hunk 12]
```diff
@@ -19,35 +20,56 @@ class Messages extends \Magento\Framework\App\Action\Action implements
...
      * @param \Oleksandrr\Chat\Model\ResourceModel\Message\CollectionFactory $messageCollectionFactory
      * @param \Psr\Log\LoggerInterface $logger
+     * @param \Magento\Customer\Model\Session $userSession
      */
     public function __construct(
         \Magento\Framework\App\Action\Context $context,
         \Oleksandrr\Chat\Model\ResourceModel\Message\CollectionFactory $messageCollectionFactory,
-        \Psr\Log\LoggerInterface $logger
+        \Psr\Log\LoggerInterface $logger,
+        \Magento\Customer\Model\Session $userSession
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/Controller/Message/Messages.php]: @pandaUbuntu Proxy is not configured here and in some other classes

File: app/code/Oleksandrr/Chat/etc/events.xml [hunk 13]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0"?>
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
+    <event name="customer_login">
+        <observer name="customer_login_observer" instance="Oleksandrr\Chat\Observer\UserLogin" />
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/etc/events.xml]: @pandaUbuntu This is frontend event, not global

File: app/code/Oleksandrr/Chat/view/frontend/templates/sidebar.phtml [hunk 14]
```diff
@@ -7,7 +7,7 @@
      }}'
 >
     <h3><?= $block->escapeHtml(__('Chat with Admin')) ?></h3>
-
+    <?= $block->getBlockHtml('formkey') ?>
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/view/frontend/templates/sidebar.phtml]: @pandaUbuntu You already get this from cookies in the JS code. Why is this needed here? Form key must not be used in the cacheable content like this - otherwise all customers will get the same form key.

File: app/code/Oleksandrr/Chat/view/frontend/web/js/sidebar.js [hunk 15]
```diff
@@ -128,19 +145,35 @@ define([
...
+             * @var {String} item.author_name
+             * @var {String} item.message
+             */
+
             $.each(list, function (index, item) {
                 var date = new Date(item.created_at);
 
                 $('#oleksandrr-chat-message-field').append(
-                    context.createMessage(item.author_name, date.toLocaleTimeString(), item.message)
+                    this.createMessage(item.author_name, date.toLocaleTimeString(), item.message)
```
Comment (maksymz) [inline app/code/Oleksandrr/Chat/view/frontend/web/js/sidebar.js]: @pandaUbuntu DOM operations are the most expensive ones. Generate content in the loop, append once. Never append in the loop unless this is really needed (most likely it is not).

**Review by maksymz — CHANGES_REQUESTED** [review]

File: app/code/Oleksandrr/Chat/Controller/Message/Messages.php (hunk 2 above)
Comment (pandaUbuntu) [reply app/code/Oleksandrr/Chat/Controller/Message/Messages.php]: Provided that it is written above in one case, this variable may not be.

File: app/code/Oleksandrr/Chat/Controller/Message/Messages.php (hunk 3 above)
Comment (pandaUbuntu) [reply app/code/Oleksandrr/Chat/Controller/Message/Messages.php]: in this case it is a native array and it has no method of count
$collection = [];

Comment (maksymz) [reply app/code/Oleksandrr/Chat/Controller/Message/Messages.php]: @pandaUbuntu You're right. And in this case naming is not really good ) Please, don't use `collection` for non-objects. Can use `$messages` or `$messageList`. By the way, this is not `Post` anyway
